### PR TITLE
fix: support retain_graph=True in warp_custom_op backward

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Fix `retain_graph=True` support in `warp_custom_op` backward pass. The Warp tape state was destructively consumed on the first backward call, causing subsequent passes through the same graph to fail with a missing-state error.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #63 

## Changes Made

- Replace `_pop_state` (destructive `.pop()`) with `_get_state` (non-destructive `.get()`) in `warp_custom_op` backward so the tape survives across multiple backward passes
- Add `state.tape.zero()` before each backward replay to prevent gradient accumulation from prior passes
- Update error message to reflect new semantics (graph freed vs state released)
- Add `retain_graph`/`create_graph` support notes to the `warp_custom_op` docstring
- Add `TestRetainGraph` class with two tests: `autograd.grad` + `backward()` pattern, and double `backward(retain_graph=True)` pattern

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

New tests:
- `TestRetainGraph::test_retain_graph_grad_then_backward` -- exercises `autograd.grad(retain_graph=True)` followed by `energy.backward()` through Ewald ops with geometry-dependent charges
- `TestRetainGraph::test_retain_graph_double_backward` -- calls `energy.backward(retain_graph=True)` twice and asserts identical gradients

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

This fix covers `retain_graph=True` without `create_graph=True`. The `create_graph=True` + `loss.backward()` pattern (second-order differentiation through Warp ops) is a separate limitation documented in the updated docstring for `warp_custom_op`. Users needing that pattern should use `hybrid_forces=True` in the electrostatics APIs, which keeps Warp ops out of the autograd graph entirely.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: test/interactions/electrostatics/bindings/torch/test_ewald.py [hunk 1]
```diff
@@ -6800,6 +6800,86 @@ def test_hybrid_reciprocal_space(self, device):
...
+        )
+        energy = energies.sum()
+        energy.backward(retain_graph=True)
+        grad_first = charges_1.grad.clone()
+
+        charges_1.grad = None
+        energy.backward()
+        grad_second = charges_1.grad.clone()
+
+        torch.testing.assert_close(grad_first, grad_second, rtol=0.0, atol=0.0)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply test/interactions/electrostatics/bindings/torch/test_ewald.py:6880]: Two tensors should have same data. Test passes.

**Review by dallasfoster — APPROVED** [review]

**Comment by dallasfoster** [comment]
/ok to test [3964f03](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/64/commits/3964f03225d7651c9b12e69915b421e715f225f8)
